dlib/media/ffmpeg_utils.h

```diff
@@ -680,26 +665,90 @@ namespace dlib
 
-            inline bool register_ffmpeg::register_library()
+            inline bool register_ffmpeg()
             {
-                avdevice_register_all();
+                static const bool REGISTERED = []
+                {
+                    avdevice_register_all();
 #if LIBAVCODEC_VERSION_INT < AV_VERSION_INT(58, 10, 100)
-                // See https://github.com/FFmpeg/FFmpeg/blob/70d25268c21cbee5f08304da95be1f647c630c15/doc/APIchanges#L91
-                avcodec_register_all();
+                    // See https://github.com/FFmpeg/FFmpeg/blob/70d25268c21cbee5f08304da95be1f647c630c15/doc/APIchanges#L91
+                    avcodec_register_all();
 #endif
 #if LIBAVFORMAT_VERSION_INT < AV_VERSION_INT(58, 9, 100) 
-                // See https://github.com/FFmpeg/FFmpeg/blob/70d25268c21cbee5f08304da95be1f647c630c15/doc/APIchanges#L86
-                av_register_all();
+                    // See https://github.com/FFmpeg/FFmpeg/blob/70d25268c21cbee5f08304da95be1f647c630c15/doc/APIchanges#L86
+                    av_register_all();
 #endif
-                return true;
-            }
+
+                    av_log_set_callback([](void* ptr, int level, const char *fmt, va_list vl) 
+                    {
+                        auto& logger = details::logger_internal_private();
+
+                        char line[256] = {0};
+                        static int print_prefix = 1;
+
+                        // Not sure if copying to vl2 is required by internal ffmpeg functions do this...
+                        va_list vl2;
+                        va_copy(vl2, vl);
+                        int size = av_log_format_line2(ptr, level, fmt, vl2, &line[0], sizeof(line), &print_prefix);
+                        va_end(vl2);
+
+                        // Remove all '\n' since dlib's logger already adds one
+                        size = std::min<int>(size, sizeof(line));
+                        line[size] = '\0';
+                        for (int i = size - 1 ; i >= 0 ; --i)
+                            if (line[i] == '\n')
+                                line[i] = ' ';
+
+                        switch(level)
+                        {
+                            case AV_LOG_PANIC:
+                            case AV_LOG_FATAL:      logger << LFATAL << line; break;
+                            case AV_LOG_ERROR:      logger << LERROR << line; break;
+                            case AV_LOG_WARNING:    logger << LWARN  << line; break;
+                            case AV_LOG_INFO:       
+                            case AV_LOG_VERBOSE:    logger << LINFO  << line; break;
+                            case AV_LOG_DEBUG:      logger << LDEBUG << line; break;
+                            case AV_LOG_TRACE:      logger << LTRACE << line; break;
+                            default: break;
+                        }
+                    });
+
+                    return true;
+                }();
+
+                return REGISTERED;
```

Comment:
I just need to statically initialize once. I don't see how I do that if I return `void`

Reply:
I mean maybe you are worried about it being optimized away, but my read on the standard is that this kind of thing, that calls other functions and does all this stuff can definitely not be optimized away, since it definitely has a ton of observable side effects.

Reply:
Does it really not work if you return void?  Like did you try?  Maybe I'm about to learn something about local static variables.  But if that really doesn't work you could definitely use `std::call_once` to do it while returning void.

Reply:
Ooooh. I have never used `std::call_once`. This could be a first. I'll try later.

Reply:
So I need a global [std::once_flag](https://en.cppreference.com/w/cpp/thread/once_flag) somewhere. I'll play with it later. I'm so used to have a function that returns a bool and some static bool somewhere being initialized with that function. It always works and i've never had a problem.

Reply:
The docs for `std::call_once()` mention this:

> Initialization of [function-local statics](https://en.cppreference.com/w/cpp/language/storage_duration#Static_local_variables) is guaranteed to occur only once even when called from multiple threads, and may be more efficient than the equivalent code using std::call_once.

Reply:
Right.  I'm pretty sure you don't need `std::call_once()` and that a local static is enough.  Like try it.  I bet it works.  My read from the standard is that you don't have to return anything here.

Reply:
You are right as usual. I always learn something from your code reviews.

Reply:
You don't even get warnings about unused variables with statics. Didn't know that.

Reply:
Hmm, maybe not. I might have to add
```
(void)REGISTERED;
```
I'll wait for tests to finish.

Reply:
Yep. Corrected. Should be good now.